server/src/main/java/org/elasticsearch/search/lookup/SourceProvider.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.lookup;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.index.fieldvisitor.LeafStoredFieldLoader;
+import org.elasticsearch.index.fieldvisitor.StoredFieldLoader;
+
+import java.io.IOException;
+
+/**
+ * Provides access to the Source of a document
+ */
+public interface SourceProvider {
+
+    /**
+     * Get the Source for the given doc within the given context
+     */
+    Source getSource(LeafReaderContext ctx, int doc) throws IOException;
+
+    /**
+     * A SourceProvider that loads source from stored fields
+     */
+    static SourceProvider fromStoredFields() {
+        StoredFieldLoader storedFieldLoader = StoredFieldLoader.sequentialSource();
+        return new SourceProvider() {
+
+            // TODO we can make this segment thread safe by keeping an array of LeafStoredFieldLoader/doc/Source
+            // records, indexed by the ordinal of the LeafReaderContext
```

Comment:
You'd want this thread safe so it can be a constant?

Reply:
One of the things stopping us doing concurrent searching across segments is that calling `source()` concurrently from different threads will screw things up.  If we have an array of per-segment loaders then we can use lucene's threaded index searcher to do searches on runtime fields, for example, because it divides things up so that each segment is only accessed by a single thread.

It can't be a constant though because we need a new one per search.

Reply:
Got it. Is there any chance we can make the source looking infrastructure take the `LeafReaderContext` as a ctor parameter? Make them per-segment objects rather than setSegment objects? That always felt cleaner to me. But it's more complex to change to if you're working with setSegment stuff.

Reply:
So add a LeafSourceProvider impl that's in between the SourceProvider and the Source? That could work, but it would take a bit of re-plumbing.  I think this PR is a good first step to it?

Reply:
Yeah. I think I'd like it better but I don't have any idea if it's worth it. Lots of replumbing, like you said.